source/grid/manifold_lib.cc

```diff
@@ -301,68 +301,146 @@ get_new_point (const std::vector<Point<spacedim> > &vertices,
 // ============================================================
 
 template <int dim, int spacedim>
-CylindricalManifold<dim,spacedim>::CylindricalManifold(const unsigned int axis,
-                                                       const double tolerance) :
-  direction (Point<spacedim>::unit_vector(axis)),
-  point_on_axis (Point<spacedim>()),
-  tolerance(tolerance)
+CylindricalManifold<dim, spacedim>::CylindricalManifold(const unsigned int axis,
+                                                        const double tolerance) :
+  ChartManifold<dim,spacedim,3>(Tensor<1,3>({0,2.*numbers::PI,0})),
+              normal_direction(compute_normal(Point<3>::unit_vector(axis))),
+              direction (Point<3>::unit_vector(axis)),
+              point_on_axis (Point<3>()),
+              tolerance(tolerance)
 {
-  Assert(spacedim > 1, ExcImpossibleInDim(1));
+  static_assert(spacedim==3,
+                "CylindricalManifold can only be used for spacedim==3!");
 }
 
 
+
 template <int dim, int spacedim>
-CylindricalManifold<dim,spacedim>::CylindricalManifold(const Point<spacedim> &direction,
-                                                       const Point<spacedim> &point_on_axis,
-                                                       const double tolerance) :
-  direction (direction),
-  point_on_axis (point_on_axis),
-  tolerance(tolerance)
+CylindricalManifold<dim, spacedim>::CylindricalManifold(const Point<spacedim> &direction_,
+                                                        const Point<spacedim> &point_on_axis_,
+                                                        const double tolerance) :
+  ChartManifold<dim,3,3>(Tensor<1,3>({0,2.*numbers::PI,0})),
+              normal_direction(compute_normal(direction_)),
+              direction (direction_/direction_.norm()),
+              point_on_axis (point_on_axis_),
+              tolerance(tolerance)
 {
-  Assert(spacedim > 2, ExcImpossibleInDim(spacedim));
+  static_assert(spacedim==3,
+                "CylindricalManifold can only be used for spacedim==3!");
 }
 
 
 
+template <int dim, int spacedim>
+Point<spacedim>
+CylindricalManifold<dim, spacedim>::compute_normal(const Tensor<1,spacedim> &vector) const
```

Comment:
Is this function similar to `compute_orthonormal_vectors()` in `vector_tools.templates.h` around line 3439? If not, maybe make them similar (e.g., take the first of the vectors computed there) and open a github issue that asks for this code to be unified somewhere.